### ISSUE TITLE
BadLocationException at org.eclipse.jface.text.link.LinkedPositionGroup.enforceDisjoint(LinkedPositionGroup.java:162)

### ISSUE THREAD

**Issue by sorincarbunaru** [eclipse-jdt/eclipse.jdt.ui#1317]
Sometimes extracting a variable and replacing all occurences either freezes, or give an exception like this:

````
org.eclipse.core.runtime.CoreException: org.eclipse.jface.text.BadLocationException
	at org.eclipse.jdt.internal.ui.text.correction.proposals.LinkedCorrectionProposal.performChange(LinkedCorrectionProposal.java:197)
	at org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal.apply(CUCorrectionProposal.java:230)
	at org.eclipse.jface.text.contentassist.CompletionProposalPopup.insertProposal(CompletionProposalPopup.java:1004)
	at org.eclipse.jface.text.contentassist.CompletionProposalPopup.insertSelectedProposalWithMask(CompletionProposalPopup.java:951)
	at org.eclipse.jface.text.contentassist.CompletionProposalPopup$3.widgetDefaultSelected(CompletionProposalPopup.java:697)
	at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:123)
	at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
	at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4273)
	at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
	at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4071)
	at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3659)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine$5.run(PartRenderingEngine.java:1155)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.e4.ui.internal.workbench.swt.PartRenderingEngine.run(PartRenderingEngine.java:1046)
	at org.eclipse.e4.ui.internal.workbench.E4Workbench.createAndRunUI(E4Workbench.java:155)
	at org.eclipse.ui.internal.Workbench.lambda$3(Workbench.java:648)
	at org.eclipse.core.databinding.observable.Realm.runWithDefault(Realm.java:342)
	at org.eclipse.ui.internal.Workbench.createAndRunWorkbench(Workbench.java:555)
	at org.eclipse.ui.PlatformUI.createAndRunWorkbench(PlatformUI.java:173)
	at org.eclipse.ui.internal.ide.application.IDEApplication.start(IDEApplication.java:152)
	at org.eclipse.equinox.internal.app.EclipseAppHandle.run(EclipseAppHandle.java:208)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.runApplication(EclipseAppLauncher.java:136)
	at org.eclipse.core.runtime.internal.adaptor.EclipseAppLauncher.start(EclipseAppLauncher.java:104)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:402)
	at org.eclipse.core.runtime.adaptor.EclipseStarter.run(EclipseStarter.java:255)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.equinox.launcher.Main.invokeFramework(Main.java:651)
	at org.eclipse.equinox.launcher.Main.basicRun(Main.java:588)
	at org.eclipse.equinox.launcher.Main.run(Main.java:1459)
Caused by: org.eclipse.jface.text.BadLocationException
	at org.eclipse.jface.text.link.LinkedPositionGroup.enforceDisjoint(LinkedPositionGroup.java:162)
	at org.eclipse.jface.text.link.LinkedPositionGroup.addPosition(LinkedPositionGroup.java:127)
	at org.eclipse.jdt.internal.ui.viewsupport.LinkedProposalModelPresenter.enterLinkedMode(LinkedProposalModelPresenter.java:100)
	at org.eclipse.jdt.internal.ui.text.correction.proposals.LinkedCorrectionProposal.performChange(LinkedCorrectionProposal.java:186)
	... 31 more
````

After trying the variable extraction operation, along with the BadLocationException, in the error log I also had a NullPointerException:

````
java.lang.NullPointerException: Cannot invoke "org.eclipse.core.filebuffers.ITextFileBuffer.getDocument()" because "this.fBuffer" is null
	at org.eclipse.ltk.core.refactoring.TextFileChange.acquireDocument(TextFileChange.java:211)
	at org.eclipse.jdt.core.refactoring.CompilationUnitChange.acquireDocument(CompilationUnitChange.java:85)
	at org.eclipse.ltk.core.refactoring.TextChange.getCurrentDocument(TextChange.java:313)
	at org.eclipse.ltk.core.refactoring.TextChange.getPreviewDocument(TextChange.java:520)
	at org.eclipse.ltk.core.refactoring.TextChange.getPreviewDocument(TextChange.java:401)
	at org.eclipse.jdt.core.manipulation.CUCorrectionProposalCore.getAdditionalProposalInfo(CUCorrectionProposalCore.java:135)
	at org.eclipse.jdt.internal.ui.text.correction.proposals.RefactoringCorrectionProposalCore.getAdditionalProposalInfo(RefactoringCorrectionProposalCore.java:77)
	at org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal.getAdditionalProposalInfo(CUCorrectionProposal.java:193)
	at org.eclipse.jface.text.contentassist.AdditionalInfoController$Timer$2$1.run(AdditionalInfoController.java:116)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
````

When Eclipse does not freeze, the result of the operation is something like this:
````
AuthorAccess authorAccess = MISSING();
AuthorAccess authorAccess = context.getAuthorAccess();
````

I have the following Eclipse:
````
Eclipse IDE for Java Developers (includes Incubating components)

Version: 2023-12 (4.30.0)
Build id: 20231201-2043
````

Version of `jdt.ui`: 	Eclipse.org	Java Development Tools UI	3.31.0.v20231115-2128	org.eclipse.jdt.ui.

An example of code that trigers a freeze is in the https://github.com/oxygenxml/oxygen-git-client-addon repo: in `ProjectMenuGitActionsProvider.doGitDiff()`, I tried to extract `pluginWS.getProjectManager()` as a variable (with the action that also replaces all occurences).

**Comment by iloveeclipse** [comment]
Is it possible to provide steps to reproduce or a minimal test case?

@jjohnstn : does stack says anything useful to you?

**Comment by jjohnstn** [comment]
@iloveeclipse The stack doesn't say much other than there is an overlapping edit generated.  Without a test, it is difficult to help debug.  There was a problem where an invalid edit may be proposed because the code wasn't cleaning up after finding errors in checkFinalConditions() but that was fixed in 2024-03.  I might suggest the user update to 2024-03 to see if that was indeed the problem.

**Comment by sorincarbunaru** [comment]
Hi,

It's rather difficult to create a minimal test case... I noticed this does not happen for every variable extraction. I didn't figure out what causes the problem.

What I can tell for sure is that, for example, it still happens for the example I provided in the initial post:

> An example of code that trigers a freeze is in the https://github.com/oxygenxml/oxygen-git-client-addon repo: in `ProjectMenuGitActionsProvider.doGitDiff()`, I tried to extract `pluginWS.getProjectManager()` as a variable (with the action that also replaces all occurences).

I guess I will have to upgrade to 2024-03, but that will be quite tedious, because I will have to install quite a few of add-ons to make sure the environment is exactly as before and exactly what I actually need for my work.

**Comment by sorincarbunaru** [comment]
I've actually just tried the scenario mentioned above with a fresh clone of that Git project and with a fresh Eclipse:
````
Eclipse IDE for Java Developers (includes Incubating components)

Version: 2024-03 (4.31.0)
Build id: 20240307-1437

Eclipse froze.

**Comment by iloveeclipse** [comment]
> Eclipse froze.

Perfect (because you can reproduce). Please create jstack thread dump and attach here.

**Comment by sorincarbunaru** [comment]
I triggered the variable extraction and then ran `jstack -l -e <PID>`. Here is the dump file.

[jstack_thread_dump.txt](https://github.com/eclipse-jdt/eclipse.jdt.ui/files/14995152/jstack_thread_dump.txt)

Let me know if it's OK.

**Comment by iloveeclipse** [comment]
Relevant threads:

```
"main" #1 prio=6 os_prio=0 cpu=10078.12ms elapsed=259.02s allocated=3142M defined_classes=13080 tid=0x00000214010c70c0 nid=0x39a8 waiting on condition  [0x000000b8ff30c000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.10/Native Method)
        - parking to wait for  <0x00000000a2ac2bb0> (a java.util.concurrent.ForkJoinTask$AdaptedCallable)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.10/LockSupport.java:341)
        at java.util.concurrent.ForkJoinTask.awaitDone(java.base@17.0.10/ForkJoinTask.java:468)
        at java.util.concurrent.ForkJoinTask.get(java.base@17.0.10/ForkJoinTask.java:979)
        at org.eclipse.jdt.internal.core.search.PatternSearchJob.performParallelSearch(PatternSearchJob.java:156)
        at org.eclipse.jdt.internal.core.search.PatternSearchJob.execute(PatternSearchJob.java:117)
        at org.eclipse.jdt.internal.core.search.processing.JobManager.performConcurrentJob(JobManager.java:384)
        at org.eclipse.jdt.internal.core.hierarchy.IndexBasedHierarchyBuilder.legacySearchAllPossibleSubTypes(IndexBasedHierarchyBuilder.java:579)
        at org.eclipse.jdt.internal.core.hierarchy.IndexBasedHierarchyBuilder.searchAllPossibleSubTypes(IndexBasedHierarchyBuilder.java:480)
        at org.eclipse.jdt.internal.core.hierarchy.IndexBasedHierarchyBuilder.determinePossibleSubTypes(IndexBasedHierarchyBuilder.java:445)
        at org.eclipse.jdt.internal.core.hierarchy.IndexBasedHierarchyBuilder.build(IndexBasedHierarchyBuilder.java:146)
        at org.eclipse.jdt.internal.core.hierarchy.TypeHierarchy.compute(TypeHierarchy.java:326)
        at org.eclipse.jdt.internal.core.hierarchy.TypeHierarchy.refresh(TypeHierarchy.java:1294)
        - locked <0x0000000088b80218> (a org.eclipse.jdt.internal.core.hierarchy.TypeHierarchy)
        at org.eclipse.jdt.internal.core.CreateTypeHierarchyOperation.executeOperation(CreateTypeHierarchyOperation.java:94)
        at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:742)
        at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:808)
        at org.eclipse.jdt.internal.core.BinaryType.newTypeHierarchy(BinaryType.java:921)
        at org.eclipse.jdt.internal.core.BinaryType.newTypeHierarchy(BinaryType.java:892)
        at org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker.findFunctionDefinition(ChangedValueChecker.java:167)
        at org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker$ReadVisitor.visit(ChangedValueChecker.java:492)
        at org.eclipse.jdt.core.dom.MethodInvocation.accept0(MethodInvocation.java:220)
        at org.eclipse.jdt.core.dom.ASTNode.accept(ASTNode.java:3300)
        at org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker.analyzeSelectedExpression(ChangedValueChecker.java:115)
        at org.eclipse.jdt.internal.corext.refactoring.util.ChangedValueChecker.<init>(ChangedValueChecker.java:93)
        at org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.evalStartAndEnd(ExtractTempRefactoring.java:1243)
        at org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.createAndInsertTempDeclaration(ExtractTempRefactoring.java:1210)
        at org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.createTempDeclaration(ExtractTempRefactoring.java:1434)
        at org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.processSelectedExpression(ExtractTempRefactoring.java:885)
        at org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.doCreateChange(ExtractTempRefactoring.java:830)
        at org.eclipse.jdt.internal.corext.refactoring.code.ExtractTempRefactoring.checkFinalConditions(ExtractTempRefactoring.java:736)
        at org.eclipse.jdt.internal.ui.text.correction.proposals.RefactoringCorrectionProposalCore.createTextChange(RefactoringCorrectionProposalCore.java:66)
        at org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal.createTextChangeViaDelegate(CUCorrectionProposal.java:288)
        at org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal.createTextChange(CUCorrectionProposal.java:255)
        - locked <0x0000000088b808e8> (a org.eclipse.jdt.internal.ui.text.correction.QuickAssistProcessor$RefactoringCorrectionProposalExtension)
        at org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal.createChange(CUCorrectionProposal.java:296)
        at org.eclipse.jdt.ui.text.java.correction.ChangeCorrectionProposal.getChange(ChangeCorrectionProposal.java:339)
        - locked <0x0000000088b808e8> (a org.eclipse.jdt.internal.ui.text.correction.QuickAssistProcessor$RefactoringCorrectionProposalExtension)
        at org.eclipse.jdt.ui.text.java.correction.ChangeCorrectionProposal.performChange(ChangeCorrectionProposal.java:132)
        at org.eclipse.jdt.internal.ui.text.correction.proposals.LinkedCorrectionProposal.performChange(LinkedCorrectionProposal.java:182)
        at org.eclipse.jdt.ui.text.java.correction.CUCorrectionProposal.apply(CUCorrectionProposal.java:228)
        at org.eclipse.jface.text.contentassist.CompletionProposalPopup.insertProposal(CompletionProposalPopup.java:1004)
        at org.eclipse.jface.text.contentassist.CompletionProposalPopup.insertSelectedProposalWithMask(CompletionProposalPopup.java:951)
        at org.eclipse.jface.text.contentassist.CompletionProposalPopup$3.widgetDefaultSelected(CompletionProposalPopup.java:697)
        at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:123)
        at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
        at org.eclipse.swt.widgets.Display.sendEvent(Display.java:4274)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1066)
        at org.eclipse.swt.widgets.Display.runDeferredEvents(Display.java:4072)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:3660)
```

```
"ForkJoinPool.commonPool-worker-4" #72 daemon prio=5 os_prio=0 cpu=2718.75ms elapsed=24.15s allocated=881M defined_classes=1 tid=0x0000021425692280 nid=0x296c runnable  [0x000000b8827fe000]
   java.lang.Thread.State: RUNNABLE
        at org.eclipse.jdt.internal.core.index.Index.query(Index.java:172)
        - parking to wait for  <0x000000008190dc90> (a java.util.concurrent.ForkJoinPool)
        at org.eclipse.jdt.internal.core.search.matching.SuperTypeReferencePattern.queryIn(SuperTypeReferencePattern.java:280)
        at org.eclipse.jdt.core.search.SearchPattern.findIndexMatches(SearchPattern.java:2517)
        at org.eclipse.jdt.core.search.SearchPattern.findIndexMatches(SearchPattern.java:2489)
        at org.eclipse.jdt.internal.core.search.matching.MatchLocator.findIndexMatches(MatchLocator.java:309)
        at org.eclipse.jdt.internal.core.search.PatternSearchJob.search(PatternSearchJob.java:240)
        at org.eclipse.jdt.internal.core.search.SubTypeSearchJob.search(SubTypeSearchJob.java:63)
        at org.eclipse.jdt.internal.core.search.PatternSearchJob.search(PatternSearchJob.java:214)
        at org.eclipse.jdt.internal.core.search.PatternSearchJob.lambda$0(PatternSearchJob.java:150)
        at org.eclipse.jdt.internal.core.search.PatternSearchJob$$Lambda$1175/0x0000000100bc0928.call(Unknown Source)
        at java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(java.base@17.0.10/ForkJoinTask.java:1428)
        at java.util.concurrent.ForkJoinTask.doExec(java.base@17.0.10/ForkJoinTask.java:373)
        at java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(java.base@17.0.10/ForkJoinPool.java:1182)
        at java.util.concurrent.ForkJoinPool.scan(java.base@17.0.10/ForkJoinPool.java:1655)
        at java.util.concurrent.ForkJoinPool.runWorker(java.base@17.0.10/ForkJoinPool.java:1622)
        at java.util.concurrent.ForkJoinWorkerThread.run(java.base@17.0.10/ForkJoinWorkerThread.java:165)

   Locked ownable synchronizers:
        - None
```

```
"Java indexing" #35 daemon prio=4 os_prio=-1 cpu=0.00ms elapsed=253.24s allocated=208B defined_classes=0 tid=0x0000021422282a80 nid=0x7d4 in Object.wait()  [0x000000b880cff000]
   java.lang.Thread.State: WAITING (on object monitor)
        at java.lang.Object.wait(java.base@17.0.10/Native Method)
        - waiting on <no object reference available>
        at java.lang.Object.wait(java.base@17.0.10/Object.java:338)
        at org.eclipse.jdt.internal.core.search.processing.JobManager.indexerLoop(JobManager.java:503)
        - locked <0x0000000082c00170> (a org.eclipse.jdt.internal.core.search.indexing.IndexManager)
        at org.eclipse.jdt.internal.core.search.processing.JobManager$$Lambda$754/0x000000010085aad8.run(Unknown Source)
        at java.lang.Thread.run(java.base@17.0.10/Thread.java:840)

   Locked ownable synchronizers:
        - None
```

@sorincarbunaru : are you sure it freezes **forever** or may be it will "unfreeze" in few minutes?
I see that a type hierarchy is created on main thread, which is not , let say, not optimal.

Seem to be added via c0e779e2349c0fe92e8961cfc8c8c4118cf6deb8  / #432

**Comment by sorincarbunaru** [comment]
Um, I guess I will have to wait forever to see if it really freezes that long... Just kidding 😆...

I waited for 20 minutes and Eclipse was still frozen.

I would recommend you try the same scenario, so that you can also reproduce the issue consistently. You just need a fresh Eclipse, and a clone of that repo. Then perform the variable extraction that I mentioned, and you should have a frozen Eclipse. 😄

**Comment by jjohnstn** [comment]
@sorincarbunaru Could you provide some information about how to set up this plug-in for building within Eclipse?  I didn't see anything in the git repo.   For example, when trying to build this project, the pom.xml file is missing the parent pom and it complains about missing oxygen-sdk jar.

**Comment by sorincarbunaru** [comment]
@jjohnstn Sure. Try again with the POM file from the attached [pom.zip](https://github.com/eclipse-jdt/eclipse.jdt.ui/files/15006370/pom.zip).

**Comment by jjohnstn** [comment]
@sorincarbunaru Thanks.   I am able to reproduce now.

**Comment by jjohnstn** [comment]
Hi @sorincarbunaru When I go to: StandAlonePluginWorkspace and attempt to do an Open Type Hierarchy, Eclipse gets 95% of the way and then freezes for a very long time; so long a time I left home for a while and when I came back hours later it had finally finished.  The final hierarchy shows multiple trees based on various interfaces that lead to the given class.  I am going to look into tailoring a hierarchy call for the extract code to use as it is only interested in subtypes of the given type.  Either way, the length of time taken to do the hierarchy needs to be investigated.  Looking at the Extract logic itself, I can make it better so as to avoid doing an unnecessary type hierarchy call when the method declaration can be found easily, but in this case, the extraction would still cause the stall because the method in question is abstract.

**Comment by jjohnstn** [comment]
@sorincarbunaru I have made a patch that removes the use of type hierarchy and replaces it with searching.  It works fast for the given scenario and addresses the freeze issue.  As mentioned, I believe that the BadLocation error should have been fixed by the recent version of Eclipse but if you have a scenario that causes the BadLocation without a freeze, please post.  If it isn't reproducible with current Eclipse or you can't remember the scenario, I would like to close this issue and you can either open a new issue or reopen this one when you are able to reproduce.

**Comment by sorincarbunaru** [comment]
@jjohnstn 

The `BadLocationException` indeed seems not to happen in Eclipse 2024-03 (4.31.0). So you can close this issue. Thanks!

I guess the fix for the freeze will most likely be available in v4.32 of Eclipse, right? Or is there a chance for some sort of a maintenance build for 4.31?

**Comment by jjohnstn** [comment]
Thanks @sorincarbunaru Yes, the fix for the freeze will be available for 4.32 M2 and is already in the I20240421-1800 I-build and nightly I-builds from now on.  We do not do maintenance builds for non-catastrophic issues as there is always an up-coming I-build repo that can be used.  If you use the I-build specified above you will be between 4.32 M1 and 4.32 M2.  If you would rather wait for a milestone, there is the 4.32 M2 which is up-coming.
